lib/services/downloads_service_backend.dart

```diff
@@ -0,0 +1,1515 @@
+          outputItems = await _jellyfinApiData.getLatestItems(
+                  includeItemTypes: "MusicAlbum", limit: 5) ??
+              [];
+        case _:
+          throw StateError("Finamp collection ${parent.id} not implemented.");
+      }
+      _downloadsService.resetConnectionErrors();
+      var stubList = outputItems
+          .map((e) => DownloadStub.fromItem(
+              item: e,
+              type: e.type == "Audio"
+                  ? DownloadItemType.song
+                  : DownloadItemType.collection))
+          .toList();
+      for (var element in stubList) {
+        _metadataCache[element.id] = Future.value(element);
+      }
+      return stubList;
+    } catch (e) {
+      _downloadsService.incrementConnectionErrors();
+      rethrow;
+    }
+  }
+
+  /// Gets the View/Library ID for the given album ID by fetching album children
+  /// of all know views.  Used by [_syncDownload] to assign libraries to items
+  /// in playlists or finampCollections.
+  Future<String?> _getAlbumViewID(String albumId) async {
+    final userHelper = GetIt.instance<FinampUserHelper>();
+    for (var view
+        in (userHelper.currentUser?.views.values ?? <BaseItemDto>[])) {
+      var children = await _getCollectionChildren(
+          DownloadStub.fromItem(type: DownloadItemType.collection, item: view));
+      var childIds =
+          children.map((e) => e.baseItem?.id).whereNotNull().toList();
+      if (childIds.contains(albumId)) {
+        return view.id;
+      }
+    }
+    return null;
+  }
+
+  /// Ensures the given node is downloaded.  Called on all required nodes with files
+  /// by [_syncDownload].  Items enqueued/downloading/failed are validated and cleaned
+  /// up before re-initiating download if needed.
+  Future<void> _initiateDownload(DownloadItem item) async {
+    switch (item.state) {
+      case DownloadItemState.complete:
+        return;
+      case DownloadItemState.notDownloaded:
+        break;
+      case DownloadItemState.enqueued: //fall through
+      case DownloadItemState.downloading:
+        if (await _downloadsService.downloadTaskQueue.validateQueued(item)) {
+          return;
+        }
+        await _downloadsService.deleteBuffer.deleteDownload(item);
+      case DownloadItemState.failed:
+      case DownloadItemState.syncFailed:
+      case DownloadItemState.needsRedownload:
+      case DownloadItemState.needsRedownloadComplete:
+        await _downloadsService.deleteBuffer.deleteDownload(item);
+    }
+
+    switch (item.type) {
+      case DownloadItemType.song:
+        return _downloadSong(item);
+      case DownloadItemType.image:
+        return _downloadImage(item);
+      case _:
+        throw StateError("???");
+    }
+  }
+
+  /// Removes unsafe characters from file names.  Used by [_downloadSong] and
+  /// [_downloadImage] for human readable download locations.
+  String? _filesystemSafe(String? unsafe) =>
+      unsafe?.replaceAll(RegExp('[/?<>\\:*|"]'), "_");
```

Comment:
Is the pipe character (|) save to use like this? I'm not sure if it's a literal match or an OR in Dart

Reply:
This is a piece of existing code from downloads_helper that got moved, so presumably it works.  It has the colon escaped but not this, which feels reasonably correct for within a regex case expression.